python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,505 @@
+
+        # see if we need to have randomly sampled entries in our mask
+        # and combinations matrices
+        self.nsamples_random = \
+            self.nsamples - self.nsamples_exact if r < self.M else 0
+
+        # we save r so we can generate random samples later
+        self.randind = r
+
+        # using numpy powerset and calculations for initial version
+        # cost is incurred only once, and generally we only generate
+        # very few samples if M is big.
+        mat, weight = powerset(self.M, r - 1, self.nsamples_exact,
+                               dtype=self.dtype)
+
+        self.mask = cp.zeros((self.nsamples, self.M), dtype=np.float32)
+        self.mask[:self.nsamples_exact] = cp.array(mat)
+
+        self.weights = cp.ones(self.nsamples, dtype=self.dtype)
+        self.weights[:self.nsamples_exact] = cp.array(weight)
+
+        self.synth_data = None
+
+        self.expected_value = self.link_fn(cp.mean(model(self.background)))
+
+        self.random_state = random_state
+
+    def explain(self,
+                X,
+                nsamples=None,
+                l1_reg='auto'):
+        shap_values = cp.zeros((1, self.M), dtype=self.dtype)
+
+        if X.ndim == 1:
+            X = X.reshape((1, self.M))
+
+        # allocating combinations array once for multiple explanations
+        if self.synth_data is None:
+            self.synth_data = cp.zeros(
+                shape=(self.N * self.nsamples, self.M),
+                dtype=np.float32,
+                order='C'
+            )
+
+        idx = 0
+        for x in X:
+            shap_values[idx] = self._explain_single_observation(
+                x.reshape(1, self.M), l1_reg
+            )
+
+        if isinstance(X, np.ndarray):
+            out_type = 'numpy'
+        else:
+            out_type = 'cupy'
+        return input_to_cuml_array(shap_values)[0].to_output(out_type)
+
+    def _explain_single_observation(self,
+                                    row,
+                                    l1_reg):
+        self.fx = cp.array(
+            model_call(X=row,
+                       model=self.model,
+                       model_gpu_based=self.model_gpu_based))
+
+        if self.nsamples_random > 0:
+            samples, self.weights[self.nsamples_exact:self.nsamples] = \
+                self._generate_number_samples_weights()
+
+        row, n_rows, n_cols, dtype = \
+            input_to_cuml_array(row, order=self.order)
+
+        cdef handle_t* handle_ = \
+            <handle_t*><size_t>self.handle.getHandle()
+        cdef uintptr_t row_ptr, bg_ptr, cmb_ptr, masked_ptr, x_ptr, smp_ptr
+
+        row_ptr = row.ptr
+        bg_ptr = self.background.__cuda_array_interface__['data'][0]
+        cmb_ptr = self.synth_data.__cuda_array_interface__['data'][0]
+        if self.nsamples_random > 0:
+            smp_ptr = samples.__cuda_array_interface__['data'][0]
+        else:
+            smp_ptr = <uintptr_t> NULL
+            maxsample = 0
```

Comment:
A lot of this could benefit from `CumlArrayDescriptor`. However, since you arent an Estimator, it might make more sense to keep these as `CumlArray` for easier access to `.ptr`